apps/prairielearn/src/question-servers/freeform.js

```diff
@@ -1806,25 +1868,35 @@ export async function grade(submission, variant, question, question_course) {
   });
 }
 
+/**
+ * @param {import('../lib/db-types.js').Variant} variant
+ * @param {import('../lib/db-types.js').Question} question
+ * @param {import('../lib/db-types.js').Course} course
+ * @param {'correct' | 'incorrect' | 'invalid'} test_type
+ * @returns {import('./types.js').QuestionServerReturnValue<import('./types.js').TestResultData>}
+ */
 export async function test(variant, question, course, test_type) {
   return instrumented('freeform.test', async () => {
     debug('test()');
     if (variant.broken_at) throw new Error('attempted to test broken variant');
 
     const context = await getContext(question, course);
+
+    /** @satisfies {import('./types.js').ExecutionData & { test_type: 'correct' | 'incorrect' | 'invalid' }} */
     let data = {
-      params: variant.params,
-      correct_answers: variant.true_answer,
+      params: variant.params ?? {},
+      correct_answers: variant.true_answer ?? {},
       format_errors: {},
       partial_scores: {},
       score: 0,
       feedback: {},
       variant_seed: parseInt(variant.variant_seed, 36),
-      options: variant.options ?? {},
+      options: { ...(variant.options ?? {}), ...getContextOptions(context) },
       raw_submitted_answers: {},
-      gradable: true,
+      gradable: /** @type {boolean} */ (true),
```

Comment:
?

Reply:
`true` has type `true`; later we need to potentially assign `false` to it so we have to deliberately widen the type. This is a consequence of using `satisfies`. Here's a minimal example: https://www.typescriptlang.org/play/?#code/C4TwDgpgBAsiBCB7RAbA8gIwFZQLxQG8oMBDAJwC5jkUISA7KAXwG4BYAKE4GNF6BnYFABmyPIWLkqwMgFdoTKPxLAAlv2GqI-WAhqYs7Lh1GIAdKTLjhJFPwgsgA

Reply:
Ah, that makes sense. Why do we want the stricter `satisfies` in this situation rather than a normal type annotation?

Reply:
Great, just making sure that we are taking advantage of the stricter types generated by `satisfies` in this situation.